commit 83af30008c8778e52f96e4f5bd2f78bc773c3d67
Author: barsnick <barsnick@users.noreply.github.com>
Date:   Mon Oct 28 09:17:56 2024 +0100

    Fix: MQTTAbstractionImpl: flush/sync MQTT Connect message immediately (#208)
    
    * Fix: MQTTAbstractionImpl: flush/sync MQTT Connect message immediately
    
    Previously, a socket was opened, but the MQTT Connect command message was
    not flushed to the broker until init() was finished. This could be critical
    if a module's init() phase waits for some reason, as the MQTT broker's
    default socket timeout might disconnect the client.
    
    This behavior was observed with the mosquitto broker, where unregistered
    sockets were consistently disconnected after 94 seconds.
    
    Insert an additional mqtt_sync() directly after mqtt_connect(), while the
    mqtt_sync() thread is not yet started.
    
    Co-authored-by: Fabian Hartung <fabian.hartung@chargebyte.com>
    Signed-off-by: Moritz Barsnick <moritz.barsnick@chargebyte.com>
    
    * Integrate early returns
    
    This commit integrates early returns in the mqtt_abstraction_impl.cpp
    file. This is done to improve the readability of the code and to reduce
    the nesting of the code.
    
    Signed-off-by: Fabian Hartung <fabian.hartung@chargebyte.com>
    
    ---------
    
    Signed-off-by: Moritz Barsnick <moritz.barsnick@chargebyte.com>
    Signed-off-by: Fabian Hartung <fabian.hartung@chargebyte.com>
    Co-authored-by: Fabian Hartung <fabian.hartung@chargebyte.com>

diff --git a/lib/mqtt_abstraction_impl.cpp b/lib/mqtt_abstraction_impl.cpp
index 52f36d0..c04bc04 100644
--- a/lib/mqtt_abstraction_impl.cpp
+++ b/lib/mqtt_abstraction_impl.cpp
@@ -484,13 +484,18 @@ bool MQTTAbstractionImpl::connectBroker(std::string& socket_path) {
     uint8_t connect_flags = MQTT_CONNECT_CLEAN_SESSION;
     /* Send connection request to the broker. */
     if (mqtt_connect(&this->mqtt_client, nullptr, nullptr, nullptr, 0, nullptr, nullptr, connect_flags,
-                     mqtt_keep_alive) == MQTT_OK) {
-        // TODO(kai): async?
-        on_mqtt_connect();
-        return true;
+                     mqtt_keep_alive) != MQTT_OK) {
+        return false;
+    }
+    // TODO(kai): async?
+    const auto error = mqtt_sync(&this->mqtt_client);
+    if (error != MQTT_OK) {
+        EVLOG_error << fmt::format("Error during MQTT sync: {}", mqtt_error_str(error));
+        return false;
     }
 
-    return false;
+    on_mqtt_connect();
+    return true;
 }
 
 bool MQTTAbstractionImpl::connectBroker(const char* host, const char* port) {
@@ -521,13 +526,18 @@ bool MQTTAbstractionImpl::connectBroker(const char* host, const char* port) {
     uint8_t connect_flags = MQTT_CONNECT_CLEAN_SESSION;
     /* Send connection request to the broker. */
     if (mqtt_connect(&this->mqtt_client, nullptr, nullptr, nullptr, 0, nullptr, nullptr, connect_flags,
-                     mqtt_keep_alive) == MQTT_OK) {
-        // TODO(kai): async?
-        on_mqtt_connect();
-        return true;
+                     mqtt_keep_alive) != MQTT_OK) {
+        return false;
+    }
+    // TODO(kai): async?
+    const auto error = mqtt_sync(&this->mqtt_client);
+    if (error != MQTT_OK) {
+        EVLOG_error << fmt::format("Error during MQTT sync: {}", mqtt_error_str(error));
+        return false;
     }
 
-    return false;
+    on_mqtt_connect();
+    return true;
 }
 
 int MQTTAbstractionImpl::open_nb_socket(const char* addr, const char* port) {
